implementation/src/main/java/io/smallrye/graphql/execution/QueryCache.java

```diff
@@ -0,0 +1,98 @@
+package io.smallrye.graphql.execution;
+
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import org.jboss.logging.Logger;
+
+import graphql.ExecutionInput;
+import graphql.execution.instrumentation.InstrumentationContext;
+import graphql.execution.instrumentation.SimpleInstrumentation;
+import graphql.execution.instrumentation.parameters.InstrumentationValidationParameters;
+import graphql.execution.preparsed.PreparsedDocumentEntry;
+import graphql.execution.preparsed.PreparsedDocumentProvider;
+import graphql.validation.ValidationError;
+
+public class QueryCache extends SimpleInstrumentation implements PreparsedDocumentProvider {
+    private static final Logger LOG = Logger.getLogger(QueryCache.class);
+    private static final int MAX_CACHE_SIZE = AccessController.doPrivileged((PrivilegedAction<Integer>) () -> {
+        return Integer.getInteger("io.smallrye.graphql.execution.queryCacheMaxSize", 2048);
+    });
+
+    private final LRUCache<String, PreparsedDocumentEntry> cache = new LRUCache<>(MAX_CACHE_SIZE);
```

Comment:
is there ever only one instance of this class?  if not, should the ThreadLocal at least be static?  Usually they are to prevent multiple ones being created.  I would recommend making it static regardless.  Not sure if the LRUCache should be static. as well.

Have you measured the size of caching the objects?  Is 2048 an ok size?  If each String and PreparedDocumentEntry is 1K in size it is only 2 MB, 10K then 20 MB, etc.

Reply:
No, there will be one instance of the cache for every GraphQL application.  For containers like Quarkus, that means that there will only be one, but for containers like Open Liberty or Wildfly, etc. there could be more than one.  I think it needs to be non-static.

Reply:
ThreadLocal should still be static.